### PR TITLE
Fixed #27670 -- Prevented shell command from masking ImportErrors.

### PR DESCRIPTION
**Pull request by inglesp**
[Ticket](https://code.djangoproject.com/ticket/27670)

### CODE REVIEW

File: tests/shell/tests.py [hunk 1]
```diff
@@ -2,10 +2,24 @@
...
 
 from django import __version__
-from django.core.management import call_command
+from django.core.management import CommandError, call_command
 from django.test import SimpleTestCase, mock
 from django.test.utils import captured_stdin, captured_stdout, patch_logger
 
+try:
+    import IPython
+    del IPython
```
Comment (adamchainz) [inline tests/shell/tests.py]: I don't see why these `del`s are needed - just a vague worry about leaving them in the module namespace?

File: django/core/exceptions.py [hunk 2]
```diff
@@ -192,3 +192,8 @@ def __repr__(self):
 class EmptyResultSet(Exception):
     """A database query predicate is impossible."""
     pass
+
+
+class NoSuchShell(Exception):
```
Comment (adamchainz) [inline django/core/exceptions.py]: I don't know if this needs to be a core exception, it seems too specific to the `shell` command to me

Comment (timgraham) [inline django/core/exceptions.py]: Agreed -- putting that in `shell.py` should be fine. Maybe that's better to inherit from `ImportError` and call it `ShellImportError`?

File: django/core/exceptions.py [hunk 3]
```diff
@@ -192,3 +192,8 @@ def __repr__(self):
...
+    """A shell does not exist."""
```
Comment (timgraham) [inline django/core/exceptions.py]: does not exist -> isn't installed?

File: tests/shell/tests.py (hunk 1 above)
Comment (timgraham) [inline tests/shell/tests.py]: The usual pattern is 
```
try:
    import IPython
except ImportError:
   IPython = None
```
No need for an extra variable.

File: django/core/management/commands/shell.py [hunk 4]
```diff
@@ -127,11 +130,17 @@ def handle(self, **options):
...
+        if options['interface']:
+            shell = options['interface']
             try:
                 return getattr(self, shell)(options)
-            except ImportError:
-                pass
-        raise ImportError("Couldn't load any of the specified interfaces.")
+            except NoSuchShell:
+                raise CommandError("Couldn't load {} interface.".format(shell))
+
```
Comment (timgraham) [inline django/core/management/commands/shell.py]: chop blank line

File: tests/shell/tests.py [hunk 5]
```diff
@@ -29,3 +43,13 @@ def test_stdin_read(self, select):
             stdin.seek(0)
             call_command('shell')
         self.assertEqual(stdout.getvalue().strip(), '100')
+
+    @unittest.skipIf(ipython_installed, 'IPython is installed')
+    def test_shell_with_ipython_not_installed(self):
+        with self.assertRaises(CommandError):
```
Comment (timgraham) [inline tests/shell/tests.py]: Please use `assertRaisesMessage()`.

File: django/core/management/commands/shell.py [hunk 6]
```diff
@@ -127,11 +130,17 @@ def handle(self, **options):
...
+        else:
+            for shell in self.shells:
+                try:
+                    return getattr(self, shell)(options)
+                except NoSuchShell:
+                    pass
+            raise CommandError("Couldn't load any interface.")
```
Comment (timgraham) [inline django/core/management/commands/shell.py]: I think it's impossible to reach this since the `python` shell doesn't raise `NoSuchShell`.

File: django/core/management/commands/shell.py [hunk 7]
```diff
@@ -127,11 +130,17 @@ def handle(self, **options):
...
-        for shell in available_shells:
+        if options['interface']:
+            shell = options['interface']
             try:
                 return getattr(self, shell)(options)
-            except ImportError:
-                pass
-        raise ImportError("Couldn't load any of the specified interfaces.")
+            except NoSuchShell:
+                raise CommandError("Couldn't load {} interface.".format(shell))
```
Comment (timgraham) [inline django/core/management/commands/shell.py]: Maybe "Couldn't import <shell>." would be slightly better wording.

File: django/core/management/commands/shell.py [hunk 8]
```diff
@@ -64,11 +65,13 @@ def ipython(self, options):
                 pass
             else:
                 return
-        # no IPython, raise ImportError
-        raise ImportError("No IPython")
+        raise NoSuchShell("No IPython")
```
Comment (knbk) [inline django/core/management/commands/shell.py]: This still hides any `ImportError` if it's raised during the startup of IPython. I think you need to raise `NoSuchShell` when importing specific IPython versions fails, but before trying to initialize the shell, and let any other errors during initialization (including `ImportError`) bubble up.

**Comment by inglesp** [comment]
Thanks @adamchainz, @timgraham, and @knbk for your reviews.  I think I've addressed all the comments.

However, there's a problem with the new tests, in that they pass when run individually or as part of the full test suite with `--parallel=1`, but as soon as tests run in parallel, the tests fail.  This is because something (I haven't worked out what yet) is closing stdin, causing the `select` [here](https://github.com/inglesp/django/blob/09823383190665f5cffa8dd0a9fc85f3a4c17b22/django/core/management/commands/shell.py#L142) to return something, which means we never get to the new code.

Is there a way for a test to detect whether it's being run in parallel?  If so, could we skip these tests in that case?

File: django/core/management/commands/shell.py [hunk 9]
```diff
@@ -39,36 +44,47 @@ def add_arguments(self, parser):
 
     def _ipython_pre_011(self):
         """Start IPython pre-0.11"""
-        from IPython.Shell import IPShell
+        try:
+            from IPython.Shell import IPShell
+        except ImportError:
+            raise ShellImportError
```
Comment (adamchainz) [inline django/core/management/commands/shell.py]: I think it's preferable to raise instances rather than classes. Python 3 will instantiate it for you when required anyway ([docs](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement)):

> Otherwise, raise evaluates the first expression as the exception object. It must be either a subclass or an instance of BaseException. If it is a class, the exception instance will be obtained when needed by instantiating the class with no arguments.

**Comment by adamchainz** [comment]
> If so, could we skip these tests in that case?

I think we should just fix the fundamental problem. It was probably only added 14 days ago in #7681 , maybe not detected then because there weren't any other tests running `shell` after ones it added.

**Comment by inglesp** [comment]
Would changing

```
 if sys.platform != "win32" and select.select([sys.stdin], [], [], 0)[0]:
    exec(sys.stdin.read())
    return
```

to

```
if sys.platform != "win32" and select.select([sys.stdin], [], [], 0)[0]:
    data = sys.stdin.read()
    if data:
        exec(data)
        return
```

be an adequate fix?

Or should we aim to fix the problem of stdin getting closed when tests are run in parallel?

**Comment by adamchainz** [comment]
I think your change is against the spirit of #7681 , which was trying to bring `./manage.py shell` in line with `python`, which still "executes" an empty `stdin` and exits:

```sh
$ python <<< ''
$
```

That said it's such an edge case it might be more convenient to just do it. Fixing the parallel test environment to match the non-parallel one is likely to prevent future similar bugs with `stdin` of course.

**Comment by inglesp** [comment]
The root cause of the problem is that running tests in parallel uses the multiprocessing module which [closes the sys.stdin for subprocesses and replaces it with os.devnull](https://github.com/python/cpython/blob/master/Lib/multiprocessing/util.py#L364-L381).  Since [reading from /dev/null always returns EOF](http://man7.org/linux/man-pages/man4/null.4.html), when we run our tests in parallel,`select` will always show that `sys.stdin` is readable.

There's no way to detect whether you're reading from /dev/null or not, so I don't think this problem is fixable in the code.

Instead, I've patched `select.select()` in the tests to indicate that there's nothing to read on `sys.stdin`.  I think this is the best we can do.

File: tests/shell/tests.py [hunk 10]
```diff
@@ -29,3 +39,17 @@ def test_stdin_read(self, select):
             stdin.seek(0)
             call_command('shell')
         self.assertEqual(stdout.getvalue().strip(), '100')
+
+    @unittest.skipIf(IPython is not None, 'IPython is installed')
+    @mock.patch('django.core.management.commands.shell.select.select')
```
Comment (adamchainz) [inline tests/shell/tests.py]: Leave a comment about your finding with parallel tests so future test authors for `shell` don't have to re-learn it all. Perhaps even extract the patching into `setUp()` / `tearDown()` of the whole test class so it's automatic

Comment (inglesp) [inline tests/shell/tests.py]: Have added a comment.  Putting the patching in `setUp()` won't work since we don't want to patch for `test_stdin_read`.

**Comment by inglesp** [comment]
> Is another workaround in the line that does the `select.select` possible, checking if `sys.stdin` is pointed at `os.devnull`? Perhaps with `sys.stdin.fileno()` or similar?

This wouldn't work without a lot of hassle -- you can't easily go from a file descriptor to the name of a file.  (And `os.devnull` is just an alias for the filename `'/dev/null'` on POSIX and `'nul'` on Windows.)

You could test for `sys.stdin.fileno()` being zero, in which case it's definitely stdin, but this would cause problems where eg a user runs the command via the `subprocess` module and passes in their own `stdin`.

File: tests/shell/tests.py (hunk 10 above)
Comment (adamchainz) [inline tests/shell/tests.py]: Could unpatch in just that one test

**Comment by adamchainz** [comment]
Makes sense. Test failures all seem to be widget rendering ones

File: django/core/management/commands/shell.py (hunk 9 above)
Comment (timgraham) [inline django/core/management/commands/shell.py]: Adam, I guess you meant "classes rather than instances" since this comment isn't addressed?

Comment (adamchainz) [inline django/core/management/commands/shell.py]: I meant I thought it should change to `raise ShellImportError()` to be clearer

**Comment by timgraham** [comment]
I'm not sure if I like the approach in general. Would it be possible to make a change in the `python` method to raise something other than `ImportError` for the problematic case you discovered so that same boilerplate isn't needed in the other methods?

**Comment by adamchainz** [comment]
@timgraham It's not possible as far as I understand because the import errors we're trying to avoid masking occur in the non import lines e.g. `app = TerminalIPythonApp.instance()` and it's not possible in python to detect what exactly failed to import where, without doing horrible traceback inspection.

**Comment by timgraham** [comment]
Okay, Peter perhaps you can confirm that's the case -- it's not clear to me from the ticket exactly where the problem is. Anyway, that does give me the idea to first remove support for IPython < 1.0 to simplify this a bit: https://github.com/django/django/pull/7821.

**Comment by inglesp** [comment]
The original `ImportError` was actually raised when using the `python` interface [when `exec()`ing my `.pythonrc.py` file](https://github.com/django/django/blob/master/django/core/management/commands/shell.py#L80).

I should have thought through this properly sooner, but I think the fix is actually to catch any exception raised by that `exec`, print the traceback to screen, and carry on.  This would match the behaviour of the normal Python prompt, as well as that of bpython and IPython.

I'll close this PR and open two more: one to fix this problem, and one to provide a better error message when you try to use a shell that's not installed.
